### DISCUSSION TITLE
ways to simulate servo motor behavior

### DISCUSSION THREAD

**Discussion by tlobig** [godot-jolt/godot-jolt#856] [Questions]
Sadly the motor behavior in the joints does not expose a phase or current angle. So it works for things like wheels, but not in cases where you only want to drive the angle  until it hits a goal angle, like it would be with a servo.
I read in an old issue or discussion here that the work around is using the angular spring and drive the equlibrium point directly. It kinda works, but is a poor replacement when trying to simulate a robotic arm or leg. I appreciate the fact that Godot-Jolt is kind of a wrapper around Jolt and the Godot physics elements, but I would love to see robotics simulations enabled in Godot. I did raise this deliberately as a discussion as I hope to get one going for this specific topic. Anyways thank you for your efforts @mihe, so far.

**Comment by MrMinimal** [comment]
I managed to implement what you were asking for with PhysicalBoneSimulator3D and it's `modification_processed` signal. Using Generic6DOFJoint3D I link the bones together. An invisible duplicate of the skeleton used for the bones is used with CCDIK3D to calculate the target postion of the bones.
I then apply force/torque to an angular motor on Generic6DOFJoint3D, to move the bones towards their target position. The amount of torque is calculated with a PID controller script.

It's just a prototype for now, I still feel like this should be done in Godot core, not in GDScript.

https://old.reddit.com/r/godot/comments/1smxnam/combined_physical_bones_with_ik_and_pid/

There is a lot of people trying to implement active ragdolls / physical skeltons / physical animations and they would all benefit from it.
